### DISCUSSION TITLE
Is there a way to redirect from www to non-www?

### DISCUSSION THREAD

**Discussion by iggyzuk** [tokio-rs/axum#2707] [Q&A] [answered]
### Summary

I want to force non-www for my website e.g. https://www.example.com/hey -> https://example.com/hey
Is there already something like this? If not I wonder what's a good way to implement it, maybe as a middleware?
Thanks!

### axum version

0.7.4

**Comment by jplatte** [comment]
Are you talking about a case where requests to both go to the same axum server? I'd usually do this one level higher at the nginx or whatever reverse proxy you have in front that also does TLS termination.

If you don't have one at all and are receiving TLS traffic directly (e.g. via `axum-server`), you can indeed write a middleware for this. I'd recommend `axum::middleware::from_fn` with a function that checks the host header, returns a `Redirect` if it matches the www. subdomain, or runs the `Next` service otherwise.

**Reply by iggyzuk** [reply]
Exactly. I'm using fly.io, I've looked around but I could not find a way to do it.
I'll give the middleware a go, doesn't sound that difficult. Thanks!

**Comment by iggyzuk** [comment, marked as answer]
In case anyone is looking for the code.
Add it to the `Router` with: `app.layer(axum::middleware::from_fn(redirect_to_non_www))`
`AppError` is like in example: https://github.com/tokio-rs/axum/blob/main/examples/anyhow-error-response/src/main.rs

```rust
pub async fn redirect_to_non_www(request: Request, next: Next) -> Result<impl IntoResponse, AppError> {
    let host = request.headers().get("host").and_then(|h| h.to_str().ok());
    if let Some(host) = host {
        if host.starts_with("www.") {
            let uri = request.uri().clone();

            let mut parts = uri.into_parts();
            parts.scheme = Some(Scheme::HTTPS);
            parts.authority = Some(http::uri::Authority::from_str(
                host.trim_start_matches("www."),
            )?);

            let new_uri = http::Uri::from_parts(parts)?;

            let response = Response::builder()
                .status(StatusCode::MOVED_PERMANENTLY)
                .header("Location", new_uri.to_string())
                .body(Body::empty())?;

            return Ok(response);
        }
    }
    Ok(next.run(request).await)
}
```
